src/functions.date.js

```diff
@@ -108,24 +108,52 @@ for (let unit in s) {
 	}, {multiValued: true});
 }
 
-_.duration = $.extend(function(ms) {
-	var count = ms || 0;
-	var unit = "ms";
-
-	for (let nextUnit in s) {
-		var nextCount = _.msTo(nextUnit, ms);
+_.duration = $.extend(function (ms, terms) {
+	if (ms === 0 || terms === undefined) {
+		terms = 1;
+	}
 
-		if (nextCount === 0) {
-			break;
+	let timeLeft = ms || 0;
+	let ret = [];
+	let units = Object.keys(s).reverse();
+	units.push("ms");
+	let unitsUsed = [];
+	let onlyConsecutiveNonzeros = true;
+
+	units.map(function(unit) {
+		// get largest value of time unit for the remaining
+		// time to account for
+		let unitValue = unit === "ms"? timeLeft : _.msTo(unit, timeLeft);
+		timeLeft -= unitValue * (unit === "ms" ? 1 : Mavo.Functions[unit]());
+		return unitValue;
+	}).forEach(function (currentValue, index) {
+		if (currentValue !== 0 && ret.length < terms) {
+			let unitProperPlurality = currentValue === 1 && units[index] !== "ms" ? units[index].slice(0, -1) : units[index];
+			ret.push(currentValue + " " + Mavo.Functions.phrase($this, unitProperPlurality));
+
+			// so we have a list of unit terms used to later determine whether they're consecutive
+			unitsUsed.push(units[index]);
+		};
+	});
+
+	ret.filter(function (currentValue, index) {
```

Comment:
This seems a little overengineered, unless I'm missing some complexity of the check. It seems to me that if we encounter a zero, and it's not the first term, we just stop there and don't calculate any further terms, right? If so, it should be possible to just have a single loop (possibly even a simple `for` loop with an index), and just `break;` if `i > 0 && value === 0`. Am I missing something?